remill/Arch/X86/Arch.cpp

```diff
@@ -477,35 +477,46 @@ static void DecodeImmediate(Instruction &inst,
                             xed_operand_enum_t op_name) {
   auto val = 0ULL;
   auto is_signed = false;
-  auto imm_size = xed_decoded_inst_get_immediate_width_bits(xedd);
   auto operand_size = xed_decoded_inst_get_operand_width(xedd);
 
-  CHECK(imm_size <= operand_size)
+  Operand op = {};
+  op.type = Operand::kTypeImmediate;
+  op.action = Operand::kActionRead;
+
+  if (XED_OPERAND_PTR == op_name) {
+    auto ptr_size = xed_decoded_inst_get_branch_displacement_width_bits(xedd);
+    CHECK(ptr_size <= operand_size)
+      << "Pointer size is greater than effective operand size at "
+      << std::hex << inst.pc << ".";
+    op.size = ptr_size;
+
+    val = xed_decoded_inst_get_branch_displacement(xedd);
```

Comment:
So the branch displacement here is actually a value for `PC`, and doesn't need to be added to `PC`? This function returns an `int32_t`, but `val` is a `uint64_t`. In this instance I like there to be explicit casting (via `static_cast`) so that what happens (sign extension, zero extension) is an obvious and deliberate choice by the programming, rather than implicitly added behaviour by the compiler.

Reply:
If `PC` needs to be added, then I wonder if this should be decoded as a relative branch, so the operand comes out as something like `(ADD (REG PC) IMM)`...

Reply:
@pgoodman Thanks you for the feedback.

The `PC` isn't need to be added, in case of a `jmp far` the address [is absolute and given in the operand](https://c9x.me/x86/html/file_module_x86_id_147.html). The following example taken from a real trace:

```Assembly
...
0x7715b940    64 8b 15 30 00 00 00                         mov edx, fs:[0x30]
0x7715b947    8b 92 64 04 00 00                            mov edx, [edx+0x464]
0x7715b94d    f7 c2 02 00 00 00                            test edx, 0x2
0x7715b953    74 03                                        jz 0x7715b958

0x7715b958    ea 5f b9 15 77 33 00                         jmp far 0x33:0x7715b95f

0x7715b95f    41 ff a7 f8 00 00 00                         jmp [r15+0xf8]

0x63fb1c46    49 87 e6                                     xchg r14, rsp
0x63fb1c49    45 8b 06                                     mov r8d, [r14]
0x63fb1c4c    49 83 c6 04                                  add r14, 0x4
0x63fb1c50    45 89 45 3c                                  mov [r13+0x3c], r8d
0x63fb1c54    45 89 75 48                                  mov [r13+0x48], r14d
...
```
The `jmp far` is used to switch from compatibility (i.e. `32-bit` protected) mode to `64-bit` mode. After this jump, the `PC` becomes `0x7715b95f`  and, (in the CPU context I've found that) the `cs` is updated to `0x33`.

Reply:
> This function returns an `int32_t`, but `val` is a `uint64_t`. In this instance I like there to be explicit casting (via `static_cast`) so that what happens (sign extension, zero extension) is an obvious and deliberate choice by the programming, rather than implicitly added behaviour by the compiler.

You're right, I will modify the PR.